src/inference/involution_dsl.jl

```diff
@@ -0,0 +1,440 @@
+import ForwardDiff
+import MacroTools
+import LinearAlgebra
+
+# TODO allow providing a pair of address namespaces, one for the model and one
+# for the proposal, when calling another involution function with call
+
+# TODO proposal_retval is a way that continuous data can leak unaccounted for..
+# for now, the requirement is that the return value of f_disc cannot depend on
+# continuous addresses in the model or proposal although in the future, we
+# could do AD through its return value as well using choice_gradients() and add
+# these to the Jacobian...?
```

Comment:
Is there some way we could just not expose the retval at all?  And then we can change that once we are able to handle the retval differentiably.

I think if feasible, making it impossible to do the wrong thing is preferable to saying in a doc "don't do this."

Reply:
@bzinberg The proposal retval is convenient because sometimes the proposal and the involution are walking the same data structures, and there can be reduced code duplication, (and at least, reduced unecessary code execution), if the involution can rely on some precomputation done during the proposal. Two examples of this (where the precomputation has to do with the discrete structure of the trace) are here: https://github.com/probcomp/Gen/blob/20200322-marcoct-rjmcmcjacobians/examples/gp_structure/involution_mh.jl#L137 and here: https://github.com/probcomp/Gen/blob/20200322-marcoct-rjmcmcjacobians/examples/gp_structure/involution_mh_tree.jl#L103

But I agree that it is potentially confusing. Also, I think that automating either the detection of an invalid proposal retval or its use by the involution, or automating its contribution to the Jacobian, might be possible with Zygote, but is pretty involved.

I might be okay with not allowing dependence on proposal retval, and paying the cost of some reversible jump algorithims being more verbose and/or inefficient.

@alex-lew What do you think?